### ISSUE TITLE
Use all node tokens and also entity tokens if module is installed

### ISSUE THREAD

**Issue by cptX** [backdrop-contrib/token_filter#6]
Hi, currently the token filter doesn't work with node tokens. It would be great if we could use all the node tokens plus all the entity tokens available if the entity tokens module is installed.
Also, I don't see any menu for viewing the available tokens in the "raw html" filter description. The only thing written is "Global tokens will be replaced with their respective token values (e.g. [site:name] or [current-page:title])." It would be nice to have a viewer there to see the tokens.

**Comment by bugfolder** [comment]
Hi, let's see if we can figure out what's going on.

First, can you confirm that you're using version 1.0.2 (the current release)? Older versions didn't support tokens that need context much at all.

And can you check that if you create an ordinary page node (or whatever your basic node type is), use the "Raw HTML" filter with tokens enabled, that the token [node:title] puts the title in the node?

If the answer to both of those is "yes", then can you describe in more detail your configuration where the token is not working, so I can try to reproduce the issue on my test setup?

**Comment by cptX** [comment]
Yes I'm using v1.0.2. I have tested [node:title] in a custom module in the text area using Raw HTML filter and was not working
I'll test it in a post now to see there too.
Also not to forget it is important to have access to the token window to see which tokens are available. Without it there isn't much point of having token filter.

**Comment by cptX** [comment]
In a post [node:title] worked!

**Comment by cptX** [comment]
In a post I can confirm that [node:title] works!

**Comment by bugfolder** [comment]
Yes, I'm looking into how to add the token window now. In D7, you could always see all tokens in the Help menu, but Help menu was removed in Backdrop. (You can try https://github.com/bugfolder/token_help as a stopgap.) But until I've figured how to more fully support context, having a full list won't be much use.

Knowing that you're using it in a custom module is helpful. Is it one that you wrote?

**Comment by cptX** [comment]
Sorry my mistake not custom module, I meant custom block in layout.

**Comment by cptX** [comment]
I can confirm that in a custom block in layout is not working. I tested it again. You can test it yourself too, is very simple. Add a custom block in your layout and add the [node:title] token in the text. You will see that prints nothing. Probably the connection to the current node is not populated and the token doesn't know in which node it is in

**Comment by bugfolder** [comment]
Ah, OK. Just so I understand, is this correct:

* The token appears in the Block content of the custom block, right? (Which tokens are you using?)
* The block appears on post pages, and you want the tokens to refer to the node page that the block is on

Is that right?

**Comment by cptX** [comment]
![image](https://user-images.githubusercontent.com/9040042/108223209-3ff91200-7142-11eb-9690-c8aeb9ceed44.png)

In first red box should be printed but not (it is a custom block)
The second red box has it printed and this comes from the node text.

**Comment by bugfolder** [comment]
Great, that's very helpful. (And I see you have Devel enabled; we might need to call on that.) Let me go try to reproduce this....

**Comment by cptX** [comment]
What I really need is probably more complex as I need a token that comes from the corresponding taxonomy of a forum page, so leave it for the moment to focus on the basic functionality of the basic node tokens.

**Comment by bugfolder** [comment]
Go ahead and tell me what taxonomy tokens you're using, that will still be good to know.

**Comment by bugfolder** [comment]
And I've reproduced the behavior on my test system. Give me a day or so to dig into it.

**Comment by cptX** [comment]
What I tried to achieve is to create a custom block in layout only for the forum pages and insert a token in this custom block to show the description of the corresponding taxonomy term of the forum. This is as it looks like not easy firstly because of the bug described here and secondly because it proved that the tokens I needed are not existing in core but exist in the Entity Tokens module. So if I was to achieve the above functionality the token_filter should also include the Entity Tokens tokens and I believe this will be hard. I believe that if you succeed in presenting a token browser we can check if the Entity Tokens module shows it's own tokens there too.

**Comment by bugfolder** [comment]
Looking into it further, I think a bigger issue is that since a custom block can appear on literally any type of page, there's no way for the block to "know" what node page it's appearing on (it could be appearing on the user profile, for example), so no way to pass the node as context into the token system.

**Comment by bugfolder** [comment]
Ah, maybe I spoke too soon. We can certainly detect whether we're rendering a node/% page from the original menu routing table. That might not solve your taxonomy term issue, but it would tell us what node to use as context if nothing else appears.

**Comment by cptX** [comment]
Yes I'm imagining that it will be difficult for the custom module to have this info available. Please don't focus in my specific need as I'll try to implement it differently

**Comment by bugfolder** [comment]
Understood. But now I'm curious, and it would be nice to give Token Filter broader support for tokens requiring context. So, if you don't mind, could you tell me the specific token(s) that you're putting into your custom block?

**Comment by cptX** [comment]
What I need in my custom block comes from the Entity Tokens unfortunately. I need for example the token [node:taxonomy_forums:description]
Core had only this token available for forums [node:taxonomy_forums]

**Comment by bugfolder** [comment]
Thanks. Entity Tokens seems like a natural module to support in Token Filter, if possible. I'll use that as a test case.

**Comment by cptX** [comment]
Check this out if you have time https://github.com/backdrop/backdrop-issues/issues/4952
Maybe the functionality of Entity Tokens is what the guys in Drupal tried to put in the core after the release of D7

**Comment by bugfolder** [comment]
Let me address your comment about the lack of visibility of what tokens are available, because that involves Backdrop core and its general handling of text **formats** (distinct from text **filters**).

In general, the tips for text **filters** have two forms: a short form, which is used in the expanding "Formatting options" for formatted text—and this is supposed to be a short one-liner—and a longer form. For Token Filter module, the short form is the text "Global tokens will be replaced with their respective token values (e.g. [site:name] or [current-page:title])." The long form includes the token browser, i.e., the list of all tokens (currently only global tokens).

The long form, though, is hard to find! In fact, it appears on the "Compose tips" page for the text **format**. 

* You can see all text **formats** at `/filter/tips`
* Or any one **format** at `filter/tips/<format>`, where `<format>` is the machine name of the text format (e.g., `full_html`).

So that's where you can find the list of available tokens.

How the heck you're supposed to know or find that, though, is not at all clear. Certainly I could put a link to the filter-specific "Compose tips" page in the short description of the **filter**, but the "Compose tips" page is for the entire **format**, which contains multiple **filters**. So it seems like any link to this page should be one level up: i.e., the short version of the text **format** should contain a link to its "Compose tips" page.

Obviously, this is an issue that transcends the Token Filter module. It's an issue for Backdrop core.

At any rate, I think that's a different issue from this one, which is, at root, broadening the support of context in Token Filters. I'll create a new, separate issue for the topic of providing access to the token browser, and we'll keep this one support for context.

**Comment by cptX** [comment]
Thanks for letting us know that this path /filter/tips exists, didn't know that.
I have opened this issue here firstly because I would like all the main tokens to work, secondly because there was no token browser and thirdly hoping that some extra tokens (Entity Tokens module) will get available.

Regarding the first what is the status now, did you made them work now (in the places that weren't now working)?
Now regarding the presentation why don't we present the tokens always in the way they are presented in the URL Aliases page (see following pictures)

![image](https://user-images.githubusercontent.com/9040042/108323321-323d9e00-71cf-11eb-8c19-ec0279a1abef.png)
![image](https://user-images.githubusercontent.com/9040042/108323351-3ec1f680-71cf-11eb-892a-b7d85726d087.png)

Regarding the third what are the possibilities?

**Comment by bugfolder** [comment]
> Regarding the first [because I would like all the main tokens to work] what is the status now, did you made them work now (in the places that weren't now working)?

The status now is much the same as it was yesterday: I _think_ I can expand the support for tokens that require context. It may take some time: I need to not only develop new code, but test it to make sure it doesn't break anything existing.

> Now regarding the presentation [there was no token browser] why don't we present the tokens always in the way they are presented in the URL Aliases page (see following pictures)

That is a separate issue. I created [this issue](https://github.com/backdrop-contrib/token_filter/issues/7) to address that. It's coupled to Backdrop core behavior; see [here](https://github.com/backdrop/backdrop-issues/issues/4953) and [here](https://github.com/backdrop/backdrop-issues/issues/1089) for the relevant background.

> Regarding the third [hoping that some extra tokens (Entity Tokens module) will get available] what are the possibilities?

Too soon to say. I need to learn more about what Entity Tokens does and how it does it.

Probably what will happen next is that I'll create a branch with changes, then ping here to ask you to test it.

**Comment by cptX** [comment]
OK, I'll wait for your feedback then. Thanks for the good work!
